common/djangoapps/third_party_auth/tests/specs/test_lti.py

```diff
@@ -0,0 +1,104 @@
+        (uri, _headers, body) = self.lti.sign(
+            uri=LTI_TPA_LOGIN_URL, http_method='POST',
+            headers={'Content-Type': FORM_ENCODED},
+            body={'user_id': LTI_USER_ID}
+        )
+        login_response = self.client.post(path=uri, content_type=FORM_ENCODED, data=body)
+        # The user should be redirected to the registration form
+        self.assertEqual(login_response.status_code, 302)
+        self.assertTrue(login_response['Location'].endswith(reverse('signin_user')))
+        register_response = self.client.get(login_response['Location'])
+        self.assertEqual(register_response.status_code, 200)
+        self.assertIn('currentProvider&#34;: &#34;LTI Test Tool Consumer&#34;', register_response.content)
+        self.assertIn('&#34;errorMessage&#34;: null', register_response.content)
+
+        # Now complete the form:
+        ajax_register_response = self.client.post(
+            reverse('user_api_registration'),
+            {
+                'email': EMAIL,
+                'name': 'Myself',
+                'username': EDX_USER_ID,
+                'honor_code': True,
+            }
+        )
+        self.assertEqual(ajax_register_response.status_code, 200)
+        continue_response = self.client.get(LTI_TPA_COMPLETE_URL)
+        # The user should be redirected to the dashboard
+        self.assertEqual(continue_response.status_code, 302)
+        self.assertTrue(continue_response['Location'].endswith(reverse('dashboard')))
+
+        # Now check that we can login again
+        self.client.logout()
+        self.verify_user_email(EMAIL)
+        (uri, _headers, body) = self.lti.sign(
+            uri=LTI_TPA_LOGIN_URL, http_method='POST',
+            headers={'Content-Type': FORM_ENCODED},
+            body={'user_id': LTI_USER_ID}
+        )
+        login_2_response = self.client.post(path=uri, content_type=FORM_ENCODED, data=body)
+        # The user should be redirected to the dashboard
+        self.assertEqual(login_2_response.status_code, 302)
+        self.assertEqual(login_2_response['Location'], LTI_TPA_COMPLETE_URL)
+        complete_2_response = self.client.get(login_2_response['Location'])
+        self.assertEqual(complete_2_response.status_code, 302)
+        self.assertTrue(complete_2_response['Location'].endswith(reverse('dashboard')))
+
+        # Check that the user was created correctly
+        user = User.objects.get(email=EMAIL)
+        self.assertEqual(user.username, EDX_USER_ID)
+
+    def test_reject_initiating_login(self):
+        response = self.client.get(LTI_TPA_LOGIN_URL)
+        self.assertEqual(response.status_code, 405)  # Not Allowed
+
+    def test_reject_bad_login(self):
+        response = self.client.post(
+            path=LTI_TPA_LOGIN_URL, content_type=FORM_ENCODED,
+            data="invalid=login"
+        )
+        self.assertEqual(response.status_code, 302)
+        self.assertEqual(response['Location'], 'http://testserver/')
```

Comment:
After considerable head-scratching, it turns out (quite obviously, in hindsight) that the frame-busting code reloading the page is causing the error message to be lost. I didn't get an immediate response on IRC on whether allowing this page to load in an iframe would be a bad idea, so I've set the flag to disable it. We thought up a few workarounds for preserving the error message across a reload but they're all somewhat intrusive for a template that's used everywhere.